examples/src/main/java/com/datastax/oss/driver/examples/concurrent/LimitConcurrencyRequestThrottler.java

```diff
@@ -0,0 +1,112 @@
+ */
+package com.datastax.oss.driver.examples.concurrent;
+
+import static com.datastax.oss.driver.api.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.oss.driver.api.querybuilder.QueryBuilder.insertInto;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.CqlSessionBuilder;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.internal.core.session.throttling.ConcurrencyLimitingRequestThrottler;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Creates a keyspace and tables, and loads data using Async API into them.
+ *
+ * <p>This example makes usage of a {@link CqlSession#executeAsync(String)} method, which is
+ * responsible for executing requests in a non-blocking way. It uses {@link
+ * ConcurrencyLimitingRequestThrottler} to limit number of concurrent requests to 32. It uses
+ * advanced.throttler configuration to limit async concurrency (max-concurrent-requests = 32) The
+ * max-queue-size is set to 10000 to buffer TOTAL_NUMBER_OF_INSERTS in a queue in a case of initial
+ * delay. (see application.conf)
+ *
+ * <p>Preconditions:
+ *
+ * <ul>
+ *   <li>An Apache Cassandra(R) cluster is running and accessible through the contacts points
+ *       identified by basic.contact-points (see application.conf).
+ * </ul>
+ *
+ * <p>Side effects:
+ *
+ * <ul>
+ *   <li>creates a new keyspace "examples" in the session. If a keyspace with this name already
+ *       exists, it will be reused;
+ *   <li>creates a table "examples.tbl_sample_kv". If it exists already, it will be reused;
+ *   <li>inserts {@code TOTAL_NUMBER_OF_INSERTS} rows into the table.
+ * </ul>
+ *
+ * @see <a href="http://datastax.github.io/java-driver/manual/">Java driver online manual</a>
+ */
+public class LimitConcurrencyRequestThrottler {
+  private static final int TOTAL_NUMBER_OF_INSERTS = 10_000;
+
+  public static void main(String[] args) throws InterruptedException, ExecutionException {
+
+    try (CqlSession session = new CqlSessionBuilder().build()) {
+      createSchema(session);
+      insertConcurrent(session);
+    }
+  }
+
+  private static void insertConcurrent(CqlSession session)
+      throws InterruptedException, ExecutionException {
+    PreparedStatement pst =
+        session.prepare(
+            insertInto("examples", "tbl_sample_kv")
+                .value("id", bindMarker("id"))
+                .value("value", bindMarker("value"))
+                .build());
+    // Create list of pending CompletableFutures.
+    // We will add every operation returned from executeAsync.
+    // Next, we will wait for completion of all TOTAL_NUMBER_OF_INSERTS
+    List<CompletableFuture<?>> pending = new ArrayList<>();
+
+    // For every i we will insert a record to db
+    for (int i = 0; i < TOTAL_NUMBER_OF_INSERTS; i++) {
```

Comment:
In this case, we will be queuing `TOTAL_NUMBER_OF_INSERTS` async operations, which is ok to show how the request throttler works.
It would be nice if we could create an example where few parallel async operations are launched/queued at any given time.

Reply:
I've implemented that approach as the first one:
https://github.com/datastax/java-driver/pull/1208/commits/6d89d469df51b7de4798ebd67039d2087b4c766b
But then on the driver standup, we decided to switch it to `RequestThrottler` usage. The main reason was that logic for limiting is not a simple thing and we don't want to add the burden of implementing it to the end-user. I can include both examples. wdyt?

Reply:
I think we need a more classic example of a blocking throttling mechanism using semaphores or some other form of synchronization. We've had so many requests for that feature but we can't implement it inside the driver because of its blocking nature. Therefore I think that an example illustrating how to do it in application code would be great. The basic idea is as Jorge suggests: to determine a maximum number of in-flight queries (or pending futures if you prefer) and only submit more queries when the previous ones complete.

Reply:
@adutra @jorgebay I've added a blocking example of limiting number of concurrent requests using Executor service:
https://github.com/datastax/java-driver/pull/1208/commits/9baa4fcb807546830e01f5b17df3b4519fef6873

Reply:
Note that we are queueing `TOTAL_NUMBER_OF_INSERTS` suppliers in memory initially. I'm wondering if it would it be possible to provide an example that enqueues new tasks as previous are completed.

The gist of it would be:
- Launch n parallel async executions (`n` being the concurrency level, 32).
- Once each future is completed, launch a new one (until total has been launched).

I've implementing something like that on the node.js driver, see [part1](https://github.com/datastax/nodejs-driver/blob/master/examples/concurrent-executions/execute-in-loop.js#L30-L33) and [part2](https://github.com/datastax/nodejs-driver/blob/master/examples/concurrent-executions/execute-in-loop.js#L46-L54) (being single threaded makes everything more simple, and `async`/`await` makes executing async in sequence easier, but I think the same concepts apply).

I don't want hijack this pull request for this discussion, so feel free to go forward with this awesome list of examples and then we can continue discussing this particular example on a separate thread/patch.

Reply:
Thanks for suggestions! 
I've improved the existing example with a way of limiting the number of submitted requests that are in-memory. (according to @adutra suggestion) I've used a `Semaphore` to achieve that.
I've seen your examples and tried to inspire on it a little bit. The `Async/await` in node looks awesome.

Reply:
Great! I personally still would like to see a fully asynchronous example (maybe using `thenComposeAsync()` to launch a new async execution?)  but again, it doesn't have to be in this pull request.

Reply:
> we need a more classic example of a blocking throttling mechanism using semaphores or some other form of synchronization.

I think there's a fundamental paradox here: if you're thinking of blocking or synchronization, you're not doing async anymore, it's time to switch to the driver's synchronous API.

Case in point: you don't need a semaphore in `LimitConcurrencyCustom`; that synchronization is already built into the driver, just run a synchronous `session.execute` from each executor thread.

If you want to limit the number of `session.executeAsync` without a queuing effect, you'll need some form of backpressure. Indeed what Jorge suggested is one way to do it: register a callback on each future than launches the next call. I think it's more of a generic future API example than driver-specific, but why not include an example.

Reply:
1. Concerning existing example (`LimitConcurrencyCustom`):
Yes, you right - we can change `executeAsync` to `execute`. We don't need another level of async there - the separate `executor` together with `CompletableFuture` assert that there will be 32 threads executing operations concurrently.
Concerning `Semaphore`: if we will remove it there will be a problem that @jorgebay mentioned - `we will be queueing TOTAL_NUMBER_OF_INSERTS suppliers in memory initially`.
Semaphore is used here for asserting that there is up to `IN_FLIGHT_REQUESTS = 500` of requests in memory (and effectively in the underlying executor queue).
So I think this example should change the `execute` to `executeAsync` method but leave the Semaphore. wdyt? Also, I think we can rename this example from `LimitConcurrencyCustom` to `MutliThreadingUsingSyncApi` or something similar.

2. I've created an example that leverages `thenComposeAsync` method and does not block
(of course besides waiting for all completableFutures to finish at the end of the program:
`CompletableFuture.allOf(pending.toArray(new CompletableFuture[0])).get();`)
Please take a look:
https://github.com/datastax/java-driver/pull/1208/files#diff-fed871b22bd6f7128f0216c0df4e7056